plugins/upower/fu-plugin-upower.c

```diff
@@ -39,36 +44,101 @@ fu_plugin_startup (FuPlugin *plugin, GError **error)
+					       NULL,
+					       "org.freedesktop.UPower",
+					       "/org/freedesktop/UPower/devices/DisplayDevice",
+					       "org.freedesktop.UPower.Device",
+					       NULL,
+					       error);
+	if (data->display_proxy == NULL) {
+		g_prefix_error (error, "failed to connect to upower: ");
+		return FALSE;
+	}
+
 	return TRUE;
 }
 
+static gboolean
+fu_plugin_upower_check_percentage_level (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	gdouble level;
+	guint power_type;
+	g_autoptr(GVariant) percentage_val = NULL;
+	g_autoptr(GVariant) type_val = NULL;
+
+	/* check that we "have" a battery */
+	type_val = g_dbus_proxy_get_cached_property (data->display_proxy, "Type");
+	if (type_val == NULL) {
+		g_warning ("Failed to query power type, assume AC power");
+		return TRUE;
+	}
+	power_type = g_variant_get_uint32 (type_val);
+	if (power_type != 2) {
+		g_debug ("Not running on battery (Type: %u)", power_type);
+		return TRUE;
+	}
+
+	/* check percentage high enough */
+	percentage_val = g_dbus_proxy_get_cached_property (data->display_proxy, "Percentage");
+	if (percentage_val == NULL) {
+		g_warning ("Failed to query power percentage level, assume enough charge");
+		return TRUE;
+	}
+	level = g_variant_get_double (percentage_val);
+	g_debug ("System power source is %.1f%%", level);
+
+	return level >= MINIMUM_BATTERY_PERCENTAGE;
+}
+
+static gboolean
+fu_plugin_upower_check_on_battery (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GVariant) value = NULL;
+
+	value = g_dbus_proxy_get_cached_property (data->upower_proxy, "OnBattery");
+	if (value == NULL) {
+		g_warning ("failed to get OnBattery value, assume on AC power");
+		return TRUE;
+	}
+	return g_variant_get_boolean (value);
+}
+
 gboolean
 fu_plugin_update_prepare (FuPlugin *plugin,
 			  FwupdInstallFlags flags,
 			  FuDevice *device,
 			  GError **error)
 {
-	FuPluginData *data = fu_plugin_get_data (plugin);
-	g_autoptr(GVariant) value = NULL;
+	gboolean on_battery;
+	gboolean charged_enough;
 
-	/* can we only do this on AC power */
+	/* not all devices need this */
 	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_REQUIRE_AC))
 		return TRUE;
-	value = g_dbus_proxy_get_cached_property (data->proxy, "OnBattery");
-	if (value == NULL) {
-		g_warning ("failed to get OnBattery value, assume on AC power");
-		return TRUE;
-	}
-	if (g_variant_get_boolean (value) && (flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
-		g_set_error_literal (error,
-				     FWUPD_ERROR,
-				     FWUPD_ERROR_AC_POWER_REQUIRED,
-				     "Cannot install update "
-				     "when not on AC power unless forced");
+
+	/* determine if operating on AC or battery */
+	on_battery = fu_plugin_upower_check_on_battery (plugin);
+	/* deteremine if battery high enough */
+	charged_enough = fu_plugin_upower_check_percentage_level (plugin);
+
+	if ((on_battery || !charged_enough) &&
+	   (flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
```

Comment:
Maybe two if statements would be clearer? Then at least you can do a specific error message for each case -- at the moment it's not entirely clear what the user has to do in the `!charged_enough` case.